ENH: Add DissimiliarityMatrix.from_iterable

thermokarst opened the pull request:
Moves `from_iterable` definition from `DistanceMatrix` to `DissimilarityMatrix` (`DistanceMatrix` inherits from `DissimilarityMatrix`).

Copied existing `DistanceMatrix` `from_iterable` tests.

Fixes #1343.

Thanks!

jairideout commented on CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 ## Version 0.5.0-dev  (changes since 0.5.0 go here)
 
 ### Features
+* DissimilarityMatrix now has a new constructor method called from_iterable.
```
Can you add the issue number that this fixes? Would also be nice to backtick DissimilarityMatrix and from_iterable for readability.

jairideout commented on skbio/stats/distance/_base.py:
```diff
@@ -109,6 +109,63 @@ def __init__(self, data, ids=None):
         self._id_index = self._index_list(self._ids)
 
+    @classonlymethod
+    @experimental(as_of="0.5.0")
```
change to 0.5.0-dev, we'll bump the version number at our next release

jairideout commented on skbio/stats/distance/_base.py:
```diff
@@ -109,6 +109,63 @@ def __init__(self, data, ids=None):
+            representing the distance between the two arguments.
+        key : callable or metadata key, optional
+            A function that takes one argument and returns a string
+            representing the id of the element in the distance matrix.
```
*distance* -> *dissimilarity*

thermokarst commented:
I probably need to also update the metrics used in the tests that were added (right now these are distance metrics, but I need some guidance on what to replace with). Thanks!

mortonjt commented:
Awesome!  Thanks for knocking this out @thermokarst 

For good dissimilarity metrics, I would test [pearson correlation](http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.pearsonr.html) and [spearman correlation](http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.spearmanr.html) (without the pvalues)

thermokarst commented:
@jairideout: update based on our in-person discussion today.

- Builds the `DissimilarityMatrix` using all pairs from the iterable
- Drops the hollowness check from `DissimilarityMatrix._validate`
- Adds the hollowness check to `DistanceMatrix._validate`
- Introduces a `validate` kwarg to `DissimilarityMatrix.from_iterable` (default `True`), which causes `from_iterable` to work the same as before when True, or else opt-in to the `super` implementation.
- Added some new tests to handle the `validate` kwarg
- Updated docstrings

Thanks!

jairideout commented on CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 ## Version 0.5.0-dev  (changes since 0.5.0 go here)
 
 ### Features
+* `DissimilarityMatrix` now has a new constructor method called `from_iterable`. ([#1343](https://github.com/biocore/scikit-bio/issues/1343)).
```
Can you note that `DissimilarityMatrix` now allows non-hollow matrices, and the addition of `validate=True` parameter to `DistanceMatrix.from_iterable`? Also note in the "backwards-incompatible changes (experimental)" section that `DistanceMatrix.from_iterable` default behavior is now to validate matrix by computing all pairwise distances. Instruct users to pass `validate=False` to get the previous behavior (no validation, but faster execution).

thermokarst replied:
No problem. Did I do the `validate` kwarg backwards? Right now `True` provides the previous behavior (default).

jairideout replied:
Yeah `validate` is backwards

jairideout commented on skbio/stats/distance/_base.py:
```diff
@@ -109,6 +109,57 @@ def __init__(self, data, ids=None):
+        elif keys is not None:
+            keys_ = keys
+
+        dm = np.zeros((len(iterable),) * 2)
```
Now that the entire matrix is filled in it's a bit faster to use `np.empty` instead of `np.zeros`. Can you update this call?

jairideout commented on skbio/stats/distance/_base.py:
```diff
@@ -702,6 +751,9 @@ def from_iterable(cls, iterable, metric, key=None, keys=None):
             An iterable of the same length as `iterable`. Each element will be
             used as the respective key.
+        validate : boolean, optional
+            If ``True``, assume symmetry and hollowness when constructing the
```
Can you add the following text to give the user more detail:

```
If ``True``, all pairwise distances are computed, including upper and lower triangles and the diagonal,
and the resulting matrix is validated for symmetry and hollowness. If ``False``, `metric` is assumed to
be hollow and symmetric and only the lower triangle (excluding the diagonal) is computed.
Pass ``validate=False`` if you are sure `metric` is hollow and symmetric for improved performance.
```

jairideout commented on skbio/stats/distance/_base.py:
```diff
@@ -715,28 +767,31 @@ def from_iterable(cls, iterable, metric, key=None, keys=None):
         -----
-        Symmetry and hollowness are assumed when calculating the distances via
-        `metric`. Therefore, distances are only computed for the strictly
+        When `validate=True`, distances are only computed for the strictly
```
Remove this "Notes" section since the details are covered above in the description for `validate`.

jairideout replied:
I can't comment on the lines above this diff, but can you remove "all" from the "Returns" section text? i.e. "The `metric` applied to pairwise elements in the `iterable`."

jairideout commented on skbio/stats/distance/_base.py:
```diff
@@ -715,28 +767,31 @@ def from_iterable(cls, iterable, metric, key=None, keys=None):
-        if key is not None and keys is not None:
-            raise ValueError("Cannot use both `key` and `keys` at the same"
-                             " time.")
+        if validate:
```
This `if` statement is inverted from what it should be, we want to call `super` if `validate=True`, else only compute lower triangle.

jairideout commented on skbio/stats/distance/tests/test_base.py:
```diff
@@ -113,10 +113,101 @@ def test_init_invalid_input(self):
             DissimilarityMatrix(data, [])
 
-        # Non-hollow.
-        data = [[0.0, 1.0], [1.0, 0.01]]
```
Can you add the following test cases?

- [ ] test that `DissimilarityMatrix` constructor supports non-hollow data
- [ ] test that `DissimilarityMatrix.from_iterable` supports non-hollow data
- [ ] test that `DissimilarityMatrix.from_iterable` supports asymmetric data
- [ ] test that `DistanceMatrix.from_iterable(validate=True)` is equal to `DistanceMatrix.from_iterable(validate=False)` when passed a valid (i.e., symmetric and hollow) metric
- [ ] test that `DistanceMatrix.from_iterable(validate=False)` "works" with a non-hollow/asymmetric metric (i.e. doesn't raise an error)

jairideout commented:
Thanks @thermokarst, done reviewing!

thermokarst commented:
Thanks @jairideout! I think I have addressed all of your comments.

jairideout commented on CHANGELOG.md:
```diff
@@ -3,8 +3,12 @@
+* `DistanceMatrix.from_iterable` now accepts a `validate=True` parameter. ([#1343](https://github.com/biocore/scikit-bio/issues/1343)).
 
 ### Backward-incompatible changes [stable]
+* `DistanceMatrix.from_iterable` default behavior is now to validate matrix by computing all pairwise distances. Pass `validate=False` to get the previous behavior (no validation, but faster execution).([#1343](https://github.com/biocore/scikit-bio/issues/1343)).
```
move this under "experimental"

jairideout commented:
Thanks @thermokarst!!
